ffi_internals/src/parsing/field_attributes.rs

```diff
@@ -0,0 +1,117 @@
+use syn::{Attribute, Ident, Lit, Meta, NestedMeta, Path};
+
+/// Field-level FFI helper attributes.
+///
+#[derive(Debug, Clone)]
+pub struct FieldAttributes {
+    /// If `Some`, a path to the type that this field should be exposed as. This type must meet
+    /// some prerequisites:
+    /// 1. It must be FFI-safe (either because it's a primitive value or derives its own FFI with
+    /// `ffi_derive`).
+    /// 1. It must have a `From<T> for U` impl, where `T` is the native type of the field and `U` is
+    /// the type referenced by the `expose_as` `Path`.
+    ///
+    /// This is necessary for exposing remote types where we want to derive an FFI, but don't
+    /// control the declaration of the type.
+    ///
+    pub expose_as: Option<Path>,
```

Comment:
I wonder if there'd be any benefit from making a type that enforces these constraints?

```rust
trait ExposeAs<T>: From<T> {
    fn expose_as(&self) -> Option<Path>;
}

struct ProxyObject<T, U> {
    native_type: std::marker::PhantomData<T>,
    rust_type: std::marker::PhantomData<U>,
}

impl<T, U> ExposeAs for ProxyObject<T, U>
where
    U: From<T>
{
    fn expose_as(&self) -> Option<Path> {
        todo!("I dunno, Nick, you tell me");
    }
}

// Your existing thing...
pub struct FieldAttributes<T, U> {
    proxy_object: ProxyObject<T, U>,
    // etc.
}

Reply:
Ooooooh, that's an excellent idea 🥂  Gonna try putting something like that together.

Reply:
Aww I don't think that's actually possible because this is all proc macro code. 😞 

When this compiles, all we have is the syntax fragment; there's no type information available about whatever exists at `syn::Path`, so we can't prove that when the type at that path compiles, it's going to implement any particular traits (nor can we provide types for `T` and `U` at this stage, since the appropriate types for those parameters don't exist when `FieldAttributes` compiles).

The _derived_ code will fail to compile because it relies on that `From` impl, but we won't know that until the compiler tries to compile the code produced by the proc macros (i.e., after the proc macros are done).

Unless I'm missing something that would get this working?

Reply:
> The _derived_ code will fail to compile because it relies on that `From` impl

Ah ok, I missed that the first time around. That seems reasonable; from the developer's perspective, as long as compilation of their crate fails, I think that's what matters.